### PR TITLE
fix #6319 - Sort trackables in inventory by name or tracking code

### LINKED ISSUE #6319: Sort trackables on log page by code rather than by name
Received by a user:
> [...]would it be possible to list the trackable inventory by tracking code? Right now, it is listed in alphabetical order with the code below it. The only issue, especially when you move many trackables like myself, most of the time the user doesn't know the name of the trackable, only knows the code. So putting the codes alphabetically would be amazing and much faster. 

I never thought about that, but I fully agree. If I put trackables into the cache I always look for its tracking number and sorting by that would indeed make this more easy.
Maybe we could also put it into the first row (bigger font) and the name into the second row (smaller font) or make both lines with same font size.

### PR DESCRIPTION
**Pull request by FreddieSanchez**
This attempts to resolve #6319. I'd like some feedback on this. It feels clunky.

I still need to address the formatting as @Lineflyer suggested.

> Maybe we could also put it into the first row (bigger font) and the name into the second row (smaller font) or make both lines with same font size.

### CODE REVIEW

**Comment by rsudev** [comment]
ok to test

**Comment by pstorch** [comment]
It works so far.
I would suggest to remember the sort order in the users preferences (Settings).
I'm thinking about the `Sort By` button: do we find a better place for it? There are already lots of buttons on this page. Maybe put it into the menu? Other opinions?

**Comment by Lineflyer** [comment]
My opinion... No button at all and sort by code plus have the code in first row to make clear its sorted by that.

**Comment by FreddieSanchez** [comment]
Saving the sort method in the setting makes sense. It preserves the setting across all caches (keeps you from having to select the same sorting function for each cache). Moving the button to the menu also makes sense. It will mimic the functionality of the GeoCache sort  in the NearBy activity screen.

@Lineflyer's approach is by far the easiest to implement. Maybe I can just default the sort to the code.

**Comment by Lineflyer** [comment]
If there is no button but just a menu item I am fine as well...my proposal intended to not clutter the view with more buttons.

**Comment by FreddieSanchez** [comment]
OK - thanks for the review. I'll make the suggested changes tonight/tomorrow and push the updates to this branch.

**Comment by FreddieSanchez** [comment]
Updated the PR. Please take another look. 

The only piece missing is saving the selection in the settings.

**Comment by pstorch** [comment]
"Sort by" in the menu instead of the extra button: ok.
Now Code and Name are in one row. It's more compact this way. If it is ok, @Lineflyer has to say.
It looks like the order is wrong (Ascending vs. Descending?).
If sorting by code, then view a trackables details and going back to the log activity, the sort order is resetted (should be kept).

File: main/src/cgeo/geocaching/log/LogCacheActivity.java [hunk 1]
```diff
@@ -227,12 +236,14 @@ public ViewHolder(final View rowView) {
...
 
     private void updateTrackablesList() {
+
         final TrackableLog[] trackablesArray = getSortedTrackables().toArray(new TrackableLog[trackables.size()]);
         final ListView inventoryList = ButterKnife.findById(this, R.id.inventory);
         inventoryList.setAdapter(new TrackableLogAdapter(this, R.layout.logcache_trackable_item, trackablesArray));
         ViewUtils.setListViewHeightBasedOnItems(inventoryList);
 
-        ButterKnife.findById(this, R.id.inventory_box).setVisibility(trackables.isEmpty() ? View.GONE : View.VISIBLE);
+        final LinearLayout inventoryBox = ButterKnife.findById(this, R.id.inventory_box);
```
Comment (pstorch) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: Why introducing a temp variable here?

File: main/src/cgeo/geocaching/log/LogCacheActivity.java [hunk 2]
```diff
@@ -227,12 +236,14 @@ public ViewHolder(final View rowView) {
     }
 
     private void updateTrackablesList() {
+
```
Comment (pstorch) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: Remove blank line at start of the method

File: main/src/cgeo/geocaching/log/LogCacheActivity.java [hunk 3]
```diff
@@ -247,15 +258,55 @@ public void onClick(final View view) {
         });
     }
 
-    private ArrayList<TrackableLog> getSortedTrackables() {
-        final ArrayList<TrackableLog> sortedTrackables = new ArrayList<>(trackables);
-        Collections.sort(sortedTrackables, new Comparator<TrackableLog>() {
+    private void selectTrackableSort() {
+
```
Comment (pstorch) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: Please remove the extra blank line

File: main/src/cgeo/geocaching/log/LogCacheActivity.java [hunk 4]
```diff
@@ -247,15 +258,55 @@ public void onClick(final View view) {
...
+        comparators.add(new Comparator<TrackableLog>() {
+
+            @Override
+            public int compare(final TrackableLog lhs, final TrackableLog rhs) {
+                return TextUtils.COLLATOR.compare(lhs.trackCode, rhs.trackCode);
+            }
+        });
+
+
+
```
Comment (pstorch) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: Collapse the blank lines to one

File: main/src/cgeo/geocaching/log/LogCacheActivity.java [hunk 5]
```diff
@@ -247,15 +258,55 @@ public void onClick(final View view) {
...
+                trackableSort = comparators.get(position);
+                trackableSortPositionSelected = position;
+                updateTrackablesList();
+                dialog.dismiss();
+            }
+        });
+        alert.create().show();
+    }
+
+
```
Comment (pstorch) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: same here, collapse blank lines into one

**Comment by pstorch** [comment]
@FreddieSanchez In the code I only had some minor comments. Please squash your commits into one when you send the final one.

**Comment by SammysHP** [comment]
@pstorch You can use the "start review" function of GitHub for such comments. That will track all comments in a review that is listed at the end of the PR with the status.

**Comment by pstorch** [comment]
@SammysHP OK, will use it next time

File: main/src/cgeo/geocaching/log/LogCacheActivity.java (hunk 1 above)
Comment (FreddieSanchez) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: This was left over from adding the button. I'll remove it.

File: main/src/cgeo/geocaching/log/LogCacheActivity.java (hunk 5 above)
Comment (FreddieSanchez) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: Agreed. Will remove.

File: main/src/cgeo/geocaching/log/LogCacheActivity.java (hunk 4 above)
Comment (FreddieSanchez) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: Agreed. Will remove.

File: main/src/cgeo/geocaching/log/LogCacheActivity.java (hunk 3 above)
Comment (FreddieSanchez) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: Agreed. Will remove.

File: main/src/cgeo/geocaching/log/LogCacheActivity.java (hunk 2 above)
Comment (FreddieSanchez) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: Agreed. Will remove.

File: main/src/cgeo/geocaching/settings/Settings.java [hunk 6]
```diff
@@ -1032,8 +1033,26 @@ public static boolean getIncludeFoundStatus() {
     public static void setIncludeFoundStatus(final boolean includeFoundStatus) {
         putBoolean(R.string.pref_includefoundstatus, includeFoundStatus);
     }
+    /**
+     * Get Trackable inventory sort method based on the last Trackable inventory sort method.
+     *
+     * @return
+     *          The last Trackable Action or RETRIEVED_IT
```
Comment (pstorch) [inline main/src/cgeo/geocaching/settings/Settings.java]: Copy & Paste error ;)

File: main/src/cgeo/geocaching/log/LogCacheActivity.java [hunk 7]
```diff
@@ -247,15 +248,48 @@ public void onClick(final View view) {
         });
     }
 
-    private ArrayList<TrackableLog> getSortedTrackables() {
-        final ArrayList<TrackableLog> sortedTrackables = new ArrayList<>(trackables);
-        Collections.sort(sortedTrackables, new Comparator<TrackableLog>() {
+    private void setTrackableSort(int position) {
```
Comment (pstorch) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: add final modifier

File: main/src/cgeo/geocaching/log/LogCacheActivity.java [hunk 8]
```diff
@@ -247,15 +248,48 @@ public void onClick(final View view) {
...
+        final List<Comparator<TrackableLog>> comparators = new ArrayList<>();
```
Comment (pstorch) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: Create a List of Comparators on each call of this setter? Maybe you find another way to do it.

Comment (FreddieSanchez) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: Agreed. Moved this to static members of the class.

File: main/src/cgeo/geocaching/log/LogCacheActivity.java (hunk 7 above)
Comment (FreddieSanchez) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: Done.

File: main/src/cgeo/geocaching/settings/Settings.java (hunk 6 above)
Comment (FreddieSanchez) [inline main/src/cgeo/geocaching/settings/Settings.java]: I should stop using copy & paste... All it does is lead to problems.

**Comment by FreddieSanchez** [comment]
I think that last commit resolves all of @pstorch's comments and suggestions. However, it seems to have upset the codacy bot.

**Comment by rsudev** [comment]
I think the remark of the codacy bot is a valid one. @pstorch suggested to move the comparator list from the setter method to somewhere else. From your comment I would deduce that you intended to create a static initializer for this collection (which seems feasible to me, as it is defined compile-time and never changes), but created an instance-initializer instead, which runs on every instantiation of the class.

File: main/src/cgeo/geocaching/log/LogCacheActivity.java [hunk 9]
```diff
@@ -85,6 +85,44 @@
...
+            @Override
+            public int compare(final TrackableLog lhs,  final TrackableLog rhs) {
+                return TextUtils.COLLATOR.compare(lhs.name, rhs.name);
+            }
+        });
+
+        private final int position;
+        private final int label;
+        private final Comparator<TrackableLog> comparator;
+        public static TrackableComparator[] comparators = new TrackableComparator[]
```
Comment (pstorch) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: I think this public mutable array can be omitted. 
Instead add a static method `TrackableComparator.findByPosition(final int position)`

File: main/src/cgeo/geocaching/log/LogCacheActivity.java [hunk 10]
```diff
@@ -247,15 +285,30 @@ public void onClick(final View view) {
...
     }
 
-    private ArrayList<TrackableLog> getSortedTrackables() {
-        final ArrayList<TrackableLog> sortedTrackables = new ArrayList<>(trackables);
-        Collections.sort(sortedTrackables, new Comparator<TrackableLog>() {
+    private void selectTrackableSort() {
+        final Builder alert = new AlertDialog.Builder(this);
+        alert.setTitle(res.getString(R.string.log_tb_sortby));
+        
+        final String[] tbSortLabels = new String[TrackableComparator.comparators.length];
```
Comment (pstorch) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: `TrackableComparator.values().length`

File: main/src/cgeo/geocaching/log/LogCacheActivity.java [hunk 11]
```diff
@@ -247,15 +285,30 @@ public void onClick(final View view) {
...
+        for (final TrackableComparator tc: TrackableComparator.comparators) {
```
Comment (pstorch) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: `TrackableComparator.values()`

File: main/src/cgeo/geocaching/log/LogCacheActivity.java [hunk 12]
```diff
@@ -247,15 +285,30 @@ public void onClick(final View view) {
...
+                dialog.dismiss();
             }
         });
+        alert.create().show();
+    }
+
+    private ArrayList<TrackableLog> getSortedTrackables() {
+        final int position = Settings.getTrackableInventorySortMethod();
+        final ArrayList<TrackableLog> sortedTrackables = new ArrayList<>(trackables);
+        Collections.sort(sortedTrackables, TrackableComparator.comparators[position].getComparator());
```
Comment (pstorch) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: `TrackableComparator.findByPosition(position).getComparator()`

File: main/res/layout/logcache_trackable_item.xml [hunk 13]
```diff
@@ -73,4 +75,4 @@
         tools:text="action"
         android:maxLines="1" />
 
-</LinearLayout>
```
Comment (UniQP) [inline main/res/layout/logcache_trackable_item.xml]: These whitespace fixes (3 occurrences) are unrelated and should be removed from the commits (you can open a separate PR to tackle them).

**Comment by UniQP** [comment]
For the final version, you should also merge the fixup commits into the commit they fix (it's nice to have them separately for review though).

File: main/res/layout/logcache_trackable_item.xml (hunk 13 above)
Comment (SammysHP) [inline main/res/layout/logcache_trackable_item.xml]: These are added automatically by many editors. IMHO this is ok.

Comment (UniQP) [inline main/res/layout/logcache_trackable_item.xml]: It's orthogonal to the idea of the PR/commit, and IMHO it then should be handled separately.

File: main/src/cgeo/geocaching/log/LogCacheActivity.java (hunk 9 above)
Comment (FreddieSanchez) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: Didn't realize there was a ```.values()``` method. Will update the PR to use this.

File: main/src/cgeo/geocaching/log/LogCacheActivity.java (hunk 10 above)
Comment (FreddieSanchez) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: Done.

File: main/src/cgeo/geocaching/log/LogCacheActivity.java (hunk 11 above)
Comment (FreddieSanchez) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: Done.

**Pull request merged**